### DISCUSSION TITLE
Repeater's `extraItemActions` method could accept a Closure and pass the current item into it

### DISCUSSION THREAD

**Discussion by MtDalPizzol** [filamentphp/filament#10413] [Ideas]
I'm trying to add an extra action to repeater items.

- In the repeater I have only some fields related to each item that can be quickly edited.
- The action opens a form to edit additional info about the item.
- I'd like to set the model for the action's form, since I'm dealing with some relationships.

Here's what I have:

```php
Repeater::make('contentsMedia')
  ->extraItemActions([ // Only accepts array =(
    \Filament\Forms\Components\Actions\Action::make('edit')
      ->fillForm(function (array $arguments, Repeater $component) {
        return $component->getRawItemState($arguments['item']); // <------ I can get the current item's data here
      })
      ->form(function (Form $form) {  // <------ Can't inject item's data here
        return $form
          ->model(MISSING_ITEM_MODEL) // <------ I'd like to set the model for the form here, being the current item's model
          ->schema([
            // ... so I can use fields with relationships here
          ]);
    })
    ->action(function (array $data): void {
      // Can only access form data here, not the form instance =(
    }),
  ])
```
Here's what would be better:

```php
Repeater::make('contentsMedia')
  ->extraItemActions(function (array $data, Model $itemRecord) { // Accepts a Closure that receives the current item's data and Eloquent instance if suitable
    return [
      \Filament\Forms\Components\Actions\Action::make('edit')
        ->fillForm(function (array $arguments, Repeater $component) {
          return $component->getRawItemState($arguments['item']); // <------ I can get the current item's data here
        })
        ->form(function (Form $form) use ($itemRecord) {  // <------ Inject the item's model here
          return $form
            ->model($itemRecord) // <------ Could use the current item's model here
            ->schema([
              // ... use fields with relationships here
            ]);
      })
      ->action(function (array $data, Form $form): void { // Inject the form instance here
        $form->saveRelationships(); // This would be ideal =)
      }),
  ])
```

So, the point is that, currently, I couldn't find a way to set the model for the form in the scenario of a repeater. We are left with 2 options that are not ideal IMO:

1. Manually handle all the form data in the `->action` method, which seems a waste, since we could use the built in Filament awesomeness for simple jobs like that.

2. Create a separate component for that form and handle it's open/closed state and current item being edited via events, which also seems a lot of extra job not really needed if we could use the approach I outlined above.

**Comment by leandrocfe** [comment]
Did you try it using `mountUsing()`?

```php
->mountUsing(function (Form $form, Model $record) {
    $form
        ->fill(...)
        ->model($record);
})
```

**Reply by MtDalPizzol** [reply]
I was able to render and fill the form via `->mountUsing`, but I couldn't figure out how to `$form->saveRelationships()` inside the `->action()`.

Also: `$record` here isn't the model for the item, **but the model for the form that contains the repeater**. I understand that we can get the item's data via `$component->getItemState()`, but this seems like something that will be repeated a lot throughout the codebase, and having the item being directly passed into here would probably make the experience better.

**Comment by cirrion-software** [comment]
@MtDalPizzol hi did you get anywhere with this? I am working on the same type of thing but hit a brick wall

**Reply by MtDalPizzol** [reply]
Nope. Actually, I abandoned Filament and LW and got back to Inertia. It simply makes more sense for me.

**Comment by ericthanhtung** [comment]
How to add value to ToggleButtons option??

Repeater::make('members')
                                ->schema([
                                    ToggleButtons::make('answers')
                                        ->multiple()
                                        ->label('Select Answer')
                                        ->options([
                                            'A' => 'A'
                                        ])->inline(),
                                ])
                                ->extraItemActions([
                                    \Filament\Forms\Components\Actions\Action::make('addOption')
                                        ->tooltip('Add Option')
                                        ->icon('heroicon-m-plus')
                                        ->action(function (array $arguments, Repeater $component) {
                                        })
                                ])

**Comment by RichardStyles** [comment]
For future reference, there is a way to get other values associated with the item the extraItemAction in Filament v3. By using the first bit of https://filamentphp.com/docs/3.x/forms/fields/builder#adding-extra-item-actions-to-a-builder to get the item state from the builder.

Something like:
```
->form(function (array $arguments, Builder $component) {
    $state = $component->getState();

    return [
        Repeater::make('only_for_types')
            ->visible(fn() => in_array(
                $state[$arguments['item']]['type'],
                ['type-a', 'type-b']
            )
        )
    ]

```
As the action get launched with the arguments that contain the UUID, so you can use the getState of the builder to get the Builder detail for that row then use however you need.
ie. the above example the repeater in the action only shows then the builder item that you click on matches the type.
You can use the same method to add options to radio/selects depending on the builder state.

Then you can save in the action using the latter end of https://filamentphp.com/docs/3.x/forms/fields/builder#adding-extra-item-actions-to-a-builder


```
->action(array $arguments, array $data, Builder $component) {
// ensure validation made

// Update item
$state[$arguments['item']] = [
    'type' => 'type-a',
    'data' => $data,
];

// Set the new data for the builder
$component->state($state);
}
```
